File: backend_go/internal/dto/quiz.go
package dto

type QuizQuestionItem struct {
	ID            string            `json:"id"`
	Question      string            `json:"question"`
	Options       map[string]string `json:"options"`
	CorrectAnswer string            `json:"correctAnswer"`
	Explanation   string            `json:"explanation"`
	Difficulty    string            `json:"difficulty,omitempty"`
}

type QuizQuestionsResponse struct {
	Questions []QuizQuestionItem `json:"questions"`
	Total     int                `json:"total"`
}

type SubmitQuizRequest struct {
	CourseID  string             `json:"courseId" binding:"required"`
	Answers   []SubmitQuizAnswer `json:"answers" binding:"required"`
	TimeSpent int                `json:"timeSpent"`
}

type SubmitQuizAnswer struct {
	QuestionID     string `json:"questionId" binding:"required"`
	SelectedAnswer string `json:"selectedAnswer" binding:"required"`
}

type SubmitQuizResult struct {
	QuestionID string `json:"questionId"`
	IsCorrect  bool   `json:"isCorrect"`
}

type SubmitQuizResponse struct {
	Score      int                `json:"score"`
	Total      int                `json:"total"`
	Percentage int                `json:"percentage"`
	Results    []SubmitQuizResult `json:"results"`
}

File: gestory/src/types/quiz.ts
/**
 * Quiz & Game Types
 * Contracts untuk quiz dan game logic
 */

import type { QuizQuestion } from "./course";

export interface QuizSession {
  id: string;
  courseId: string;
  userId?: string;
  questions: QuizQuestion[];
  currentQuestionIndex: number;
  score: number;
  answers: QuizAnswer[];
  startedAt: Date;
  completedAt?: Date;
}

export interface QuizAnswer {
  questionId: string;
  selectedAnswer: string;
  isCorrect: boolean;
  timeTaken: number; // in seconds
}

export interface QuizSubmission {
  sessionId: string;
  answers: QuizAnswer[];
  finalScore: number;
  totalQuestions: number;
  duration: number; // in seconds
}

export interface GameQuestion {
  id: string;
  question: string;
  options: Record<string, string>;
  correctAnswer: string;
  explanation: string;
  category?: string;
  difficulty?: "easy" | "medium" | "hard";
}

export interface GameSession {
  id: string;
  userId?: string;
  questions: GameQuestion[];
  currentQuestionIndex: number;
  score: number;
  lives: number;
  startedAt: Date;
  completedAt?: Date;
}

export interface GameScore {
  sessionId: string;
  finalScore: number;
  questionsAnswered: number;
  correctAnswers: number;
  accuracy: number; // percentage
  timeSpent: number; // in seconds
  completedAt: Date;
}

export interface QuizQuestionsResponse {
  questions: GameQuestion[];
  total: number;
}
